x-pack/plugin/text-structure/src/main/java/org/elasticsearch/xpack/textstructure/transport/TransportTestGrokPatternAction.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.textstructure.transport;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.util.concurrent.EsExecutors;
+import org.elasticsearch.grok.Grok;
+import org.elasticsearch.grok.GrokBuiltinPatterns;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.textstructure.action.TestGrokPatternAction;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class TransportTestGrokPatternAction extends HandledTransportAction<TestGrokPatternAction.Request, TestGrokPatternAction.Response> {
+
+    private static final Logger logger = LogManager.getLogger(TransportTestGrokPatternAction.class);
+
+    private final ThreadPool threadPool;
+
+    @Inject
+    public TransportTestGrokPatternAction(TransportService transportService, ActionFilters actionFilters, ThreadPool threadPool) {
+        super(
+            TestGrokPatternAction.NAME,
+            transportService,
+            actionFilters,
+            TestGrokPatternAction.Request::new,
+            EsExecutors.DIRECT_EXECUTOR_SERVICE
+        );
+        this.threadPool = threadPool;
+    }
+
+    @Override
+    protected void doExecute(Task task, TestGrokPatternAction.Request request, ActionListener<TestGrokPatternAction.Response> listener) {
+        // As matching a regular expression might take a while, we run
+        // in a different thread to avoid blocking the network thread.
+        threadPool.generic().execute(() -> {
+            try {
+                listener.onResponse(getResponse(request));
+            } catch (Exception e) {
+                listener.onFailure(e);
+            }
+        });
+    }
+
+    private TestGrokPatternAction.Response getResponse(TestGrokPatternAction.Request request) {
+        Grok grok = new Grok(GrokBuiltinPatterns.get(true), request.getGrokPattern(), logger::warn);
```

Comment:
I don't think we should log errors in the Grok constructor as warnings here. It will mean that if a slightly suboptimal Grok pattern makes it into our default set then every call to this endpoint will log a warning. (This has happened in the past because, for example, an upgrade to Joni might decide to log that in `.*?` the `?` is redundant when the old version didn't.) That will create a lot of pointless log noise, because the Grok object that gets created for ingest pipelines will have already logged the same error.

Also, the [Find structure API](https://www.elastic.co/guide/en/elasticsearch/reference/current/find-structure.html) let's the user choose the ECS compatibility. So this new API should too, otherwise it won't always be possible to use it in conjunction with "Find structure". Please add a new `ecs_compatibility` option to the Request class.

```suggestion
        Grok grok = new Grok(GrokBuiltinPatterns.get(ECS_COMPATIBILITY_V1.equals(request.getEcsCompatibility())), request.getGrokPattern(), logger::debug);
```

Reply:
Should `ecs_compatibility` be a query param or a body param? And what's the general guideline for such decisions?

Reply:
I think `ecs_compatibility` should be a URL argument, as it's a simple flag.

The API design guidelines say:

1. We should not allow the same option to be passed either in the body or in the URL arguments - I know many existing APIs breach this guideline, but it's the recommended approach going forwards
2. The body should be used for arguments that could be large, or are lists (passing lists in URL arguments is painful)

Reply:
thanks